transports/quic/src/muxer.rs

```diff
@@ -0,0 +1,418 @@
+        let addr = SocketAddr::new(ip, inner.endpoint.port());
+        crate::transport::socketaddr_to_multiaddr(&addr)
+    }
+
+    pub fn remote_addr(&self) -> Multiaddr {
+        let inner = self.inner.lock();
+        let addr = inner.connection.remote_address();
+        crate::transport::socketaddr_to_multiaddr(&addr)
+    }
+
+    pub(crate) fn set_accept_incoming(&self, accept: bool) {
+        let mut inner = self.inner.lock();
+        inner.accept_incoming = accept;
+    }
+}
+
+impl<C: Crypto> StreamMuxer for QuicMuxer<C> {
+    type Substream = StreamId;
+    type OutboundSubstream = ();
+    type Error = QuicMuxerError;
+
+    fn poll_event(
+        &self,
+        cx: &mut Context,
+    ) -> Poll<Result<StreamMuxerEvent<Self::Substream>, Self::Error>> {
+        let mut inner = self.inner.lock();
+        let now = Instant::now();
+
+        while let Poll::Ready(event) = inner.endpoint.poll_channel_events(cx) {
+            inner.connection.handle_event(event);
+        }
+
+        let _max_datagrams = inner.endpoint.max_datagrams();
+        while let Some(transmit) = inner.connection.poll_transmit(now) {
+            inner.endpoint.send_transmit(transmit);
+        }
+
+        loop {
+            if let Some(timer) = inner.timer.as_mut() {
+                match Pin::new(timer).poll(cx) {
+                    Poll::Ready(expired) => {
+                        inner.connection.handle_timeout(expired);
+                        inner.timer = None;
+                    }
+                    Poll::Pending => break,
+                }
+            } else if let Some(when) = inner.connection.poll_timeout() {
+                inner.timer = Some(Timer::at(when));
+            } else {
+                break;
+            }
+        }
+
+        while let Some(event) = inner.connection.poll_endpoint_events() {
+            inner.endpoint.send_endpoint_event(event);
+        }
+
+        while let Some(event) = inner.connection.poll() {
+            match event {
+                Event::HandshakeDataReady => {}
+                Event::Connected => {
+                    // Break here so that the noise upgrade can finish.
+                    return Poll::Pending;
+                }
+                Event::ConnectionLost { reason } => {
+                    tracing::debug!("connection lost because of {}", reason);
+                    inner.substreams.clear();
+                    if let Some(waker) = inner.close_waker.take() {
+                        waker.wake();
+                    }
+                    return Poll::Ready(Err(QuicMuxerError::ConnectionLost { reason }));
+                }
+                Event::Stream(StreamEvent::Opened { dir: Dir::Bi }) => {
+                    // handled at end.
+                }
+                Event::Stream(StreamEvent::Readable { id }) => {
+                    tracing::trace!("stream readable {}", id);
+                    if let Some(substream) = inner.substreams.get_mut(&id) {
```

Comment:
Is there a scenario where one would receive a `StreamEvent::Readable` but not have a reference to it in `inner.substreams`? If not, how about an `expect` here to enforce this assumption?